### PR TITLE
[SPARK-14361][SQL]Window function exclude clause 

### PR DESCRIPTION
**Pull request by xwu0226**
#### What changes were proposed in this pull request?

The current Spark SQL does not support the `exclusion` clause, which is part of ANSI SQL2003’s `Window` syntax. For example, IBM Netezza fully supports it as shown in the [document web link](https://www.ibm.com/support/knowledgecenter/SSULQD_7.1.0/com.ibm.nz.dbu.doc/c_dbuser_window_aggregation_family_syntax.html). This PR is to fill the gap. 
##### Introduction

Below is the ANSI SQL2003’s `Window` syntax:

```
FUNCTION_NAME(expr) OVER {window_name | (window_specification)}
window_specification ::= [window_name] [partitioning] [ordering] [framing]
partitioning ::= PARTITION BY value[, value...] [COLLATE collation_name]
ordering ::= ORDER [SIBLINGS] BY rule[, rule...]
rule ::= {value | position | alias} [ASC | DESC] [NULLS {FIRST | LAST}]
framing ::= {ROWS | RANGE} {start | between} [exclusion]
start ::= {UNBOUNDED PRECEDING | unsigned-integer PRECEDING | CURRENT ROW}
between ::= BETWEEN bound AND bound
bound ::= {start | UNBOUNDED FOLLOWING | unsigned-integer FOLLOWING}
exclusion ::= {EXCLUDE CURRENT ROW | EXCLUDE GROUP | EXCLUDE TIES | EXCLUDE NO OTHERS}
```

Exclusion clause can be used to exclude certain rows from the window framing when calculating window aggregation function (e.g. AVG, SUM, MAX, MIN, COUNT, etc)  related to current row. Types of window functions that are not supported are listed below:
1. Offset functions, such as lead(), lag()
2. Ranking functions, such as rank(), dense_rank(), percent_rank(), cume_dist, ntile()
3. Row number function, such as row_number()
##### Definition

| Syntax | Description |
| --- | --- |
| EXCLUDE CURRENT ROW | Specifies excluding the current row. |
| EXCLUDE GROUP | Specifies excluding the current row and all rows that are tied with it. Ties occur when there is a match on the order column or columns. |
| EXCLUDE NO OTHERS | Specifies not excluding any rows. This value is the default if you specify no exclusion. |
| EXCLUDE TIES | Specifies excluding all rows that are tied with the current row (peer rows), but retaining the current row. |
##### Use-case Examples:
- Let's say you want to find out for every employee, where is his/her salary at compared to the average salary of those within the same department and whose ages are within 5 years younger and older. The query could be: 

``` SQL
SELECT NAME, DEPT_ID, SALARY, AGE, AVG(SALARY) AS AVG_WITHIN_5_YEAR
    OVER(PARTITION BY DEPT_ID 
         ORDER BY AGE 
         RANGE BETWEEN 5 PRECEDING AND 5 FOLLOWING 
         EXCLUDE CURRENT ROW) 
FROM EMPLOYEE 
```
- Let's say you want to compare every customer's yearly purchase with other customers' average yearly purchase who are at different age group from the current customer. The query could be: 

``` SQL
SELECT CUST_NAME, AGE, PROD_CATEGORY, YEARLY_PURCHASE, AVG(YEARLY_PURCHASE) 
    OVER(PARTITION BY PROD_CATEGORY 
         ORDER BY AGE 
         RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUND FOLLOWING 
         EXCLUDE GROUP) 
FROM CUSTOMER_PURCHASE_SUM 
```
##### Implementation

**catalyst.parser.SqlBase.g4**
- add excludeClause definition and append to each case in windowFrame
- add excludeType definition for CURRENT ROW,GOUP, TIES, NO OTHERS

**catalyst.parser.AstBuilder.scala**
- parse excludeClause to generate `ExcludeClause`object to set into `SpecifiedWindowFrame`

**catalyst.expressions.windowExpressions.scala**
- add `ExcludeClause`and `ExcludeType`
- add `ExcludeClause`object as a parameter to `SpecifiedWindowFrame`
- validate EXCLUDE GROUP/TIES requires ordered window frame

**catalyst.analysis.Analyzer.scala**
- validate exclude clause does not support `RowNumberLike`, `RankLike`and `OffsetWindowFunction`

**execution.Window.scala**
- implement the exclusion logic in sliding frame, expanding frame, shrinking frame and whole partition frame

**Some Test case examples:**
**DDL and DML:**
sql("create table table1 (col1 int, col2 int, col3 int)")
      sql("insert into table1 select 6, 12, 10")
      sql("insert into table1 select 6, 11, 4")
      sql("insert into table1 select 6, 13, 11")
      sql("insert into table1 select 6, 9, 10")
      sql("insert into table1 select 6, 15, 8")
      sql("insert into table1 select 6, 10, 1")
      sql("insert into table1 select 6, 15, 8")
      sql("insert into table1 select 6, 7, 4")
      sql("insert into table1 select 6, 7, 8")

**Queries:**

```
// sliding frame with exclude current row
select col1, col2, col3, sum(col2) 
    over (partition by col1 
       order by col3 
       rows between 2 preceding and 2 following 
       exclude current row) 
from table1 where col1 = 6
```
##### Results

| col1 | col2 | col3 | sum(col2) |
| --- | --- | --- | --- |
| 6 | 10 | 1 | 18 |
| 6 | 11 | 4 | 32 |
| **6** | **7** | **4** | **51** |
| 6 | 15 | 8 | 40 |
| 6 | 15 | 8 | 40 |
| 6 | 7 | 8 | 51 |
| 6 | 12 | 10 | 44 |
| 6 | 9 | 10 | 32 |
| 6 | 6 | 13 | 11 |

In this case, for example, if we are looking at the (6, 7, 4) row, the frame around this current row has 5 rows (6, 10, 1), (6,11, 4), **(6, 7, 4)**, (6, 15, 8), (6, 15, 8), including the current row. With the `exclude current row` clause, then the calculation of sum(col2) is on the other 4 rows, which is 10 + 11 + 15 + 15 = 51. 

```
//expanding frame with exclude current row
select col1, col2, col3, sum(col2) 
    over (partition by col1 
            order by col3 
            rows between unbounded preceding and current row 
            exclude group) 
from table1 where col1 = 6
```
##### Results

| col1 | col2 | col3 | sum(col2) |
| --- | --- | --- | --- |
| 6 | 10 | 1 | null |
| 6 | 11 | 4 | 10 |
| **6** | **7** | **4** | **10** |
| 6 | 15 | 8 | 28 |
| 6 | 15 | 8 | 28 |
| 6 | 7 | 8 | 28 |
| 6 | 12 | 10 | 65 |
| 6 | 9 | 10 | 65 |
| 6 | 13 | 11 | 86 |

In this case, for example, if we are looking at row (6, 7, 4), the frame around this current row has 3 rows (6, 10, 1), (6, 11, 4) and (6, 7, 4). Then, with `exclude group` clause, rows (6, 11, 4) and (6, 7, 4) will be excluded because the values of order-by column `col3` on these 2 rows are the same, thus in the same group. Therefore, the sum(col2) at current frame is 10. 

For testcases are located in hive/execution/**WindowFunctionExcludeSuite.scala**
#### How was this patch tested?
1. created unit tests to cover all framing cases, each with different exclude types. 
2. run regtest over catalyst, sql, and hive tests under SQL component.

### CODE REVIEW

**Comment by hvanhovell** [comment]
@xwu0226 this look interesting. I will try to get to this in the next couple of days.

The thing I am really worried about is performance. Could you please add benchmarks, comparing the old and new situations. See https://issues.apache.org/jira/secure/attachment/12745984/perf_test3.scala as a starting point.

**Comment by hvanhovell** [comment]
cc @yhuai

**Comment by xwu0226** [comment]
@hvanhovell , Thanks for the suggestions! I will look into this today..

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -1398,6 +1403,18 @@ class Analyzer(
...
   /**
+   * Exclude clause does not support
+   * RowNumberLike: ROW_NUMER()
+   * RankLike: RANK(), DENSE_RANK(), CUME_DIST(), etc.
+   * OffsetWindowFunctions: Lead(), Lag()
+   * @param wf
+   * @return
+   */
+  private def notSupportForExclude(wf: WindowFunction): Boolean =
+    wf.isInstanceOf[RowNumberLike] ||
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Check for `WindowFunction`.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 2]
```diff
@@ -713,6 +719,9 @@ PRECEDING: 'PRECEDING';
 FOLLOWING: 'FOLLOWING';
 CURRENT: 'CURRENT';
 ROW: 'ROW';
+EXCLUDE: 'EXCLUDE';
```
Comment (hvanhovell) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Are these reserved keywords? If they are not please add them to the nonReserved rule.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 3]
```diff
@@ -262,19 +270,82 @@ object SpecifiedWindowFrame {
...
+ * default behavior. Exclude types, GROUP and TIES, requires ORDER BY clause in the window
+ * specification clause.
+ * For example, users want to exclude current row from a sliding range framing:
+ * RANGE BETWEEN 2 PRECEDING AND 2 FOLLOWING EXCLUDE CURRENT ROW
+ * @param excludeType The type of exclusion as defined above
+ * @param valueOrdering The ordering operator that does the value comparison between 2 rows
+ * @param toBeCompared The projection of the orderby expression from a row
+ *
+ */
+case class ExcludeClause (
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: Do we need this class in this location? It seems like this is bound to its implementation? Just passing ExcludeTypes should be fine for now.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala [hunk 4]
```diff
@@ -721,16 +791,37 @@ private[execution] final class SlidingWindowFunctionFrame(
...
-    var i = 0
-    while (i < size) {
-      processor.update(rows.next())
-      i += 1
+    if (excludeSpec.excludeType != ExcludeNoOthers) {
+      // For the exclude cases, the content of the window frame is always changing
+      // along with the current row. For example, a row that was not included before
+      // needs to be included now, or a row that was included before needs to be excluded now.
+      // So intermediate buffer is used for now..
+      // TODO: For potential performance gains, we need to look into the AggregateProcessor
```
Comment (hvanhovell) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: There is no way this is possible. Think of a MIN aggregate for instance. What you can do is merge multiple buffers.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 2 above)
Comment (xwu0226) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: They are not reserved. I can add them to nonReserved rule. Thanks!

**Comment by yhuai** [comment]
@xwu0226 Thank you for working on this! I probably will not be able to review it before the 2.0 release. How about we revisit it later?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 3 above)
Comment (xwu0226) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: For excluding, I need an Ordering operator that can compare 2 projections that are derived out of 2 rows based on the orde-by expression. So i was thinking that in execution time, i need to construct the code for Ordering and the projection. So maybe I can put it in execution.Window.scala?

**Comment by xwu0226** [comment]
@yhuai Thank you very much for your response! I truely understand that you guys are very busy now..  Since Hermant has looked into it and provided some very helpful comments for me. Let me address the comments and you can take a look at it when you are ready.. Thanks again!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala (hunk 4 above)
Comment (xwu0226) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Window.scala]: Thanks for the suggestions! Let me try that.

**Comment by xwu0226** [comment]
@hvanhovell Per your request, I did the benchmark between master and spark-14361 branch without the exclude case. The following is the results. I will continue on the exclude cases. 

```
+-----+----------------+------+-----------+-----+
| Size|            Name|Master|SPARK-14361| Diff|
+-----+----------------+------+-----------+-----+
|    2|Entire Partition| 24093|      19106|0.793|
|    2|         Growing| 15166|      15032|0.991|
|    2|       Shrinking| 15217|      15687|1.031|
|    2|         Sliding| 16191|      16314|1.008|
|    8|Entire Partition| 14599|      14749|1.010|
|    8|         Growing| 15382|      14883|0.968|
|    8|       Shrinking| 15303|      15655|1.023|
|    8|         Sliding| 15232|      15245|1.001|
|   32|Entire Partition| 15361|      14990|0.976|
|   32|         Growing| 15922|      15073|0.947|
|   32|       Shrinking| 16851|      19129|1.135|
|   32|         Sliding| 15589|      15655|1.004|
|  128|Entire Partition| 14959|      14486|0.968|
|  128|         Growing| 14659|      15095|1.030|
|  128|       Shrinking| 18121|      18319|1.011|
|  128|         Sliding| 15495|      15030|0.970|
|  256|Entire Partition| 15373|      14951|0.973|
|  256|         Growing| 14857|      14545|0.979|
|  256|       Shrinking| 20371|      22640|1.111|
|  256|         Sliding| 15228|      14626|0.960|
| 1024|Entire Partition| 14774|      14337|0.970|
| 1024|         Growing| 15888|      15413|0.970|
| 1024|       Shrinking| 39446|      45678|1.158|
| 1024|         Sliding| 15260|      14591|0.956|
| 4096|Entire Partition| 14652|      14839|1.013|
| 4096|         Growing| 15064|      15099|1.002|
| 4096|       Shrinking|147914|     193215|1.306|
| 4096|         Sliding| 15225|      15476|1.016|
|16192|Entire Partition| 14155|      14600|1.031|
|16192|         Growing| 14040|      14498|1.033|
|16192|       Shrinking|699023|     883307|1.264|
|16192|         Sliding| 20765|      17278|0.832|
+-----+----------------+------+-----------+-----+
```
It looks like the Shrinking case gets a bit worse than master while the partition size is larger, while other cases have some minor improvements. I will look at the logic in shrinking case. I may know the answer already.

**Comment by thesuperzapper** [comment]
Is there any movement on this?

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 [hunk 5]
```diff
@@ -631,6 +631,17 @@ frameBound
...
     ;
 
+excludeClause
+    : EXCLUDE excludeType=CURRENT ROW
+    | EXCLUDE excludeType=GROUP
+    | EXCLUDE excludeType=TIES
+    | EXCLUDE excludeType=NO OTHERS
+    ;
+
+explainOption
```
Comment (viirya) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: Does this related to this change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -324,6 +324,23 @@ class Analyzer(
               aggsBuffer += e
               e
             case e if isPartOfAggregation(e) => e
+            case e: GroupingID =>
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: ditto.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 7]
```diff
@@ -1825,6 +1847,18 @@ class Analyzer(
   }
 
   /**
+   * Exclude clause does not support
+   * RowNumberLike: ROW_NUMER()
+   * RankLike: RANK(), DENSE_RANK(), CUME_DIST(), etc.
+   * OffsetWindowFunctions: Lead(), Lag()
+   * @param wf
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: I think `@param` can be removed as there is no doc for it and it's obvious. Also for `@return`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala [hunk 8]
```diff
@@ -262,19 +270,55 @@ object SpecifiedWindowFrame {
    */
   def defaultWindowFrame(
       hasOrderSpecification: Boolean,
-      acceptWindowFrame: Boolean): SpecifiedWindowFrame = {
+      acceptWindowFrame: Boolean,
+      excludeType: ExcludeType = ExcludeNoOthers): SpecifiedWindowFrame = {
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: The default window frame should be `ExcludeNoOthers`. I think we don't need to add this parameter. We can directly add `ExcludeNoOthers` when constructing `SpecifiedWindowFrame` below.

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala [hunk 9]
```diff
@@ -113,7 +132,25 @@ class WindowSpec private[sql](
...
+   *
+   * Both `start` and `end` are relative from the current row. For example, "0" means "current row",
+   * while "-1" means one off before the current row, and "5" means the five off after the
+   * current row.
+   *
+   * @param start boundary start, inclusive.
+   *              The frame is unbounded if this is the minimum long value.
+   * @param end boundary end, inclusive.
+   *            The frame is unbounded if this is the maximum long value.
+   * @since 1.4.0
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala]: 1.4.0?

File: sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala [hunk 10]
```diff
@@ -113,7 +132,25 @@ class WindowSpec private[sql](
...
+   */
+  def rangeBetweenExclude(start: Long, end: Long, exclude: String = ""): WindowSpec = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/expressions/WindowSpec.scala]: We already have `rowsBetween` which doesn't specify exclude type. So I think we don't need to have default value for `exclude`.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/WindowFunctionExcludeSuite.scala [hunk 11]
```diff
@@ -0,0 +1,1000 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive.execution
+
+import org.apache.spark.sql.{AnalysisException, QueryTest, Row}
+import org.apache.spark.sql.hive.test.TestHiveSingleton
+import org.apache.spark.sql.test.SQLTestUtils
+
+class WindowFunctionExcludeSuite extends QueryTest with SQLTestUtils with TestHiveSingleton{
```
Comment (viirya) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/WindowFunctionExcludeSuite.scala]: Any significant difference between this test suite and `window_exclude.sql`? If not, we can just have `window_exclude.sql`.

File: sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4 (hunk 5 above)
Comment (xwu0226) [inline sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBase.g4]: hmm. I don't think I introduced this change.. maybe I need to rebase to clean up my code before meaningful review. Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 7 above)
Comment (xwu0226) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: Yes. thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala (hunk 8 above)
Comment (xwu0226) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/windowExpressions.scala]: Agree.

**Comment by xwu0226** [comment]
Since there are a lot of changes by other PRs over the last year, many changes in this PR may not be applicable and I am doing the rebase and necessary rework on this PR. Will update soon.

**Comment by HyukjinKwon** [comment]
Gentle ping @xwu0226, how is the update going?

**Comment by xwu0226** [comment]
@HyukjinKwon My rebased branch has broken most of the window exclude test cases. Trying to fix.

**Comment by gatorsmile** [comment]
@xwu0226 Maybe close this PR if you do not have time to finish it? Thanks!

**Comment by xwu0226** [comment]
I will re-open this PR once I resolve the issue. Thanks! @gatorsmile

**Pull request closed without merging**